File: pset6/readability/readability.py
from cs50 import get_string
import re


# prompt user to input text
text = get_string('Text: ').split()

# initialize variables
letters = 0
words = 0
sentences = 0

# iterate over words
for word in text:
    # increment word count
    words += 1

    # count characters (letters + numbers) and increment letters
    filtered = re.sub('[^A-Za-z0-9]+', '', word)
    letters += len(filtered)

    # count sentences
    punctuations = ['.', '!', '?']
    if any(delimiter in word for delimiter in punctuations):
        sentences += 1

average = 100 / words
index = (0.0588 * letters * average) - (0.296 * sentences * average) - 15.8

# determine score
if index < 1:
    print('Before Grade 1')
elif index > 16:
    print('Grade 16+')
else:
    print(f'Grade {round(index)}')

File: pset2/readability/readability.c
#include <cs50.h>
#include <string.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
    // Prompt user to input text
    string text = get_string("Text: ");

    // initialize variables
    int char_count = 0;
    int word_count = 1;
    int sentence_count = 0;

    // count words
    for (int i = 0; i < strlen(text); i += 1)
    {
        // add to character count
        if ((text[i] >= 'a' && text[i] <= 'z') || (text[i] >= 'A' && text[i] <= 'Z'))
        {
            char_count += 1;
        }
        // add to sentence count
        else if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentence_count += 1;
        }
        // add to word count
        else if (text[i] == ' ')
        {
            word_count += 1;
        }
    }

    /*
    Test the word / letter / sentence count:
    printf("Chars: %i\nWords: %i\nSentences: %i\n", char_count, word_count, sentence_count);
    */

    float grade = 0.0588 * (100 * (float) char_count / word_count) - 0.296 * (100 * (float) sentence_count / word_count) - 15.8;

    // return grade depending on text score
    if (grade < 16 && grade >= 0)
    {
        printf("Grade %i\n", (int) round(grade));
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Before Grade 1\n");
    }
}